// @opentelemetry/resource-detector-azure/build/esm/types.js
/*
 * Copyright The OpenTelemetry Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
export var AZURE_APP_SERVICE_STAMP_RESOURCE_ATTRIBUTE = 'azure.app.service.stamp';
export var CLOUD_RESOURCE_ID_RESOURCE_ATTRIBUTE = 'cloud.resource_id';
export var REGION_NAME = 'REGION_NAME';
export var WEBSITE_HOME_STAMPNAME = 'WEBSITE_HOME_STAMPNAME';
export var WEBSITE_HOSTNAME = 'WEBSITE_HOSTNAME';
export var WEBSITE_INSTANCE_ID = 'WEBSITE_INSTANCE_ID';
export var WEBSITE_OWNER_NAME = 'WEBSITE_OWNER_NAME';
export var WEBSITE_RESOURCE_GROUP = 'WEBSITE_RESOURCE_GROUP';
export var WEBSITE_SITE_NAME = 'WEBSITE_SITE_NAME';
export var WEBSITE_SLOT_NAME = 'WEBSITE_SLOT_NAME';
export var WEBSITE_SKU = 'WEBSITE_SKU';
export var FUNCTIONS_VERSION = 'FUNCTIONS_EXTENSION_VERSION';
export var FUNCTIONS_MEM_LIMIT = 'WEBSITE_MEMORY_LIMIT_MB';
export var AZURE_VM_METADATA_HOST = '169.254.169.254';
export var AZURE_VM_METADATA_PATH = '/metadata/instance/compute?api-version=2021-12-13&format=json';
export var AZURE_VM_SCALE_SET_NAME_ATTRIBUTE = 'azure.vm.scaleset.name';
export var AZURE_VM_SKU_ATTRIBUTE = 'azure.vm.sku';
